### PR TITLE
Feature/conn

### PR DESCRIPTION
**Pull request by Kevin-Chen0**
## Proposed changes

Added the following:

- HTTP connection and config: /packages/packages/fetchai/*
- Pet Store YAML file example: /examples/http_ex/*

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

The HTTP connection.py is still largely a WIP with only the baseline functions.

### CODE REVIEW

File: packages/fetchai/connections/http/connection.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                )
+                raise ValueError("Cannot send message.")
+
+        self._httpCall.send_message(
+            sender_address=envelope.sender,
+            receiver_address=envelope.to,
+            protocol=str(envelope.protocol_id),
+            context=b"None",
+            payload=envelope.message,
+        )
```
Comment (Totoual) [inline packages/fetchai/connections/http/connection.py]: Keep in mind that if you are using the p2p library from fetch is specific for the p2p implementation that we have in the Ledger constelation. The message in the library is in the form of :
``` 
 msg = {FROM: {NODE_ADDRESS: self.node_address,
                      SENDER_ADDRESS: send_address},
               TO: {NODE_ADDRESS: self.node_address,
                    RECEIVER_ADDRESS: receive_address},
               PROTOCOL: protocol,
               CONTEXT: context,
               PAYLOAD: payload}
```
I mean, I am not sure if this library is useful for what you are doing :)

File: examples/http_ex/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# HTTP agent example
```
Comment (DavidMinarsch) [inline examples/http_ex/README.md]: All documentation should be in the docs folder. Please remove here.

File: packages/fetchai/connections/http/__init__.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2020 Fetch.AI Limited
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/__init__.py]: Please make it 2018 - 2020

File: packages/fetchai/connections/http/connection.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+name: http
+author: fetchai
+version: 0.1.0
+license: Apache-2.0
+fingerprint: ""
+description: "The HTTP connection that wraps a RESTful API specification."
+class_name: HTTPConnection
+protocols: [http]
+restricted_to_protocols: ["fetchai/http:0.1.0"]
+excluded_protocols: ["fetchai/http:0.1.0"]
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.yaml]: this should be empty: `excluded_protocols: []`

File: packages/fetchai/connections/http/connection.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+name: http
+author: fetchai
+version: 0.1.0
+license: Apache-2.0
+fingerprint: ""
+description: "The HTTP connection that wraps a RESTful API specification."
+class_name: HTTPConnection
+protocols: [http]
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.yaml]: this should be `protocols: ["fetchai/http:0.1.0"]`

File: packages/fetchai/connections/http/connection.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+restricted_to_protocols: ["fetchai/http:0.1.0"]
+excluded_protocols: ["fetchai/http:0.1.0"]
+config:
+  addr: ${addr:127.0.0.1}
+  port: ${port:8000}
+  api: '' # put here the dotted path to your API spec name.
+dependencies:
+  api: {}
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.yaml]: please remove `api: {}`, it should be `dependencies: {}`

File: packages/fetchai/connections/http/connection.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                self.stopped = True
+                self.thread.join()
+
+
+class HTTPConnection(Connection):
+    """Proxy to the functionality of the RESTful API."""
+
+    def __init__(self,
+                 address: Address,
+                 api_spec: str,  # In JSON serialized form, might switch to yaml or server_stub later
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: The swagger spec above is in yaml, so why not go with a path to that file: `api_spec_path:`

File: packages/fetchai/connections/http/connection.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+                 provider_addr: str,
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: what are `provider_addr` and `provider_port` used for?

File: packages/fetchai/connections/http/connection.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+                # self._httpCall.disconnect()
+                self.stopped = True
+                self.thread.join()
+
+
+class HTTPConnection(Connection):
+    """Proxy to the functionality of the RESTful API."""
+
+    def __init__(self,
+                 address: Address,
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: what is address used for?

File: packages/fetchai/connections/http/connection.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+                 api_spec: str,  # In JSON serialized form, might switch to yaml or server_stub later
+                 provider_addr: str,
+                 provider_port: int = 10000,
+                 *args,
+                 **kwargs
+                 ):
+        """
+        Initialize a connection to an RESTful API.
+
+        :param address: the address used in the protocols.
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: Not matching the function signature

File: packages/fetchai/connections/http/connection.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        """
+
+        # the following api_spec format checks will be in their own function check_api(api_spec)
+        try:
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: We should load a yaml file (see here for an example: https://github.com/fetchai/agents-aea/blob/master/aea/configurations/loader.py#L81)

File: packages/fetchai/connections/http/connection.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+        Get the HTTP connection from the connection configuration.
+
+        :param address: the address of the agent.
+        :param connection_configuration: the connection configuration object.
+            :addr - RESTful API address
+            :port - RESTful API port
+            :api_spec - configuation of setting-up a RESTful-to-envelope messaging with given skill or other core component
+        :return: the connection object
+        """
+        addr = cast(str, connection_configuration.config.get("addr"))
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: see above

File: packages/fetchai/connections/http/connection.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        provider_addr: str,
+        provider_port: int,
+        excluded_protocols: Optional[Set[PublicId]] = None,
+    ):
+        """
+        Initialize a channel.
+
+        :param address: the address
+        """
+        self.address = address
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: note, in the P2PChannel, the P2PChannel acts as the client. Here, the HTTPChannel acts as a server.

File: packages/fetchai/connections/http/connection.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+
+    def connect(self):
+        """
+        Connect.
+
+        :return: an asynchronous queue, that constitutes the communication channel.
+        """
+        with self.lock:
+            if self.stopped:
+                self._httpCall = HTTPCalls(
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: not relevant, as this is supposed to be the server

Comment (DavidMinarsch) [reply packages/fetchai/connections/http/connection.py]: to start with, you could make HTTPChannel inherit from https://docs.python.org/2/library/basehttpserver.html#BaseHTTPServer.BaseHTTPRequestHandler
and here an example https://wiki.python.org/moin/BaseHttpServer. Then you just need to implement the `do_..` methods. i.e. do_GET, do_HEAD, do POST and create the envelopes which you hand off to the Connection

Comment (DavidMinarsch) [reply packages/fetchai/connections/http/connection.py]: this isn't ready for production but it is ok for the demo

File: packages/fetchai/connections/http/connection.py (hunk 1 above)
Comment (DavidMinarsch) [reply packages/fetchai/connections/http/connection.py]: Agreed, the P2P connection has very little relevance here

File: packages/fetchai/connections/http/connection.py (hunk 14 above)
Comment (Kevin-Chen0) [reply packages/fetchai/connections/http/connection.py]: Thanks for that, I will check out BaseHTTPServer.

**Comment by DavidMinarsch** [comment]
1. Yes please move `examples/http_ex/README.md`. In examples we just keep readmes for things not worth mentioning in the docs. So `gym` example can stay there.
2. agreed, would change to host and port.
3. we should only add functionality we know we need. Atm we don't seem to need it so let's remove it.

**Comment by Kevin-Chen0** [comment]
**Updates:**
1) Fixed the /http/connection.yaml file.
2) Updated /http_ex/apis/petstore.yaml file to latest OpenAPI v3.0.2 one found here: http://petstore.swagger.io:8080/.
3) Added the /http_ex/apis/petstore_sim.yaml file, which is simplified version of petstore.yaml used for HTTPConntection testing. 
4) Wrote two functions in /http/connection.py that convert the paths in api_spec into Envelopes each. Although the API responses in JSON format is the Envelope message atm, that may change on how the Envelope receives handles it. The full API path is specified in the uri of the Envelope.
5) Added HTTP Server and Handler in connection.py that replaces httpCalls from P2P. 

**Next steps:**
1) Continue populating the HTTPHandler.
2) Begin drafting the _petstoresim_ Skill using the petstore_sim.yaml.
3) Write the send/receiving Envelop functions in conntection.py between AEA and skill. This is for when the front end client is making requests.
4) Test using CLI tool, such as between two different computers. One for front-end client and other for HTTP server AEA.
5) Update the docs.

File: packages/fetchai/connections/http/connection.py [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+                # self._httpCall.disconnect()
+                # self.stopped = True
+                # self.thread.join()
+
+
+class HTTPConnection(Connection):
+    """Proxy to the functionality of the web RESTful API."""
+
+    def __init__(self,
+                 # address: Address,
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: Please remove all unnecessary codes, comments should be used sparingly.

File: packages/fetchai/connections/http/connection.py [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+                 api_path: str,  # Directory path of the API YAML file.
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: Please provide this description in the docstring of the parameter or make the parameter name more explicit.

File: packages/fetchai/connections/http/connection.py [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+            with open(api_path, 'r') as f:
+                api_spec = yaml.safe_load(f)
+        except FileNotFoundError:
+            logger.error("API specification YAML source file not found. Please double-check filename and path.")
+            return
+
+        if kwargs.get("connection_id") is None:
+            kwargs["connection_id"] = PublicId("fetchai", "http", "0.1.0")
+        super().__init__(*args, **kwargs)
+        # self.address = address
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: see above

File: packages/fetchai/connections/http/connection.py [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+
+        :param connection_configuration: the connection configuration object.
+            :host - RESTful API hostname / IP address
+            :port - RESTful API port number
+            :api - Directory path and filename of the API spec YAML source file.
+        :return: the connection object
+        """
+        addr = cast(str, connection_configuration.config.get("host"))
+        port = cast(int, connection_configuration.config.get("port"))
+        api_path = cast(str, connection_configuration.config.get("api"))
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: Please use same var name in config if possible

File: packages/fetchai/connections/http/connection.py [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+
+        restricted_to_protocols_names = {
+            p.name for p in connection_configuration.restricted_to_protocols
+        }
+        excluded_protocols_names = {
+            p.name for p in connection_configuration.excluded_protocols
+        }
+        return HTTPConnection(
+            # address,
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: see above

File: packages/fetchai/connections/http/connection.py [hunk 20]
```diff
@@ -0,0 +1,359 @@
...
+    def _decode_path(self, path_name: str, path_value: dict, separator: str = '/'):
+        """Path --> Envelope
+
+        Convert specified API path name and value into in_queue'd Envelope.
+        Envelope's msg consists of the response options for each REST req in path.
+        """
+        path_split = path_name.split(separator)
+
+        to = path_split[1].strip()  # to: name of Skill (petstoresim)?
+        sender = 'TBD'  # hardcoded for now
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: One way would be to create a uuid for each request. Then the sender is the uuid. You send the envelope into the agent, whilst blocking on the response to the request. Once a matching envelope with the same uuid as received comes back from within the agent (or a timeout is hit) your serve the response.

File: packages/fetchai/connections/http/connection.py [hunk 21]
```diff
@@ -0,0 +1,359 @@
...
+        protocol_id = PublicId.from_str('fetchai/http:0.1.0')  # hardcoded for now
+        uri = URI(f"http://{self.host}:{self.port}{path_name}")
+        context = EnvelopeContext(connection_id=self._connection_id, uri=uri)
+        # Msg consists of all of the responses for each REST req in JSON format.
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: not sure about this

File: packages/fetchai/connections/http/connection.py [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
+        except OSError:
+            logger.error(f"{host}:{port} is already in use, please try another Socket.")
+        # Process API specifications into Envelopes to be sent to InBox
+        self._process_api(api_spec)
+        logger.info("Initialized the HTTP Server.")
+
+    def _process_api(self, api_spec):
+        """Process the api_spec.
+
+        Decode all paths to get envelops each, and put it in the agent's inbox.
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: not sure why you;d want to do that.

File: packages/fetchai/connections/http/connection.py [hunk 23]
```diff
@@ -0,0 +1,359 @@
...
+
+    def do_HEAD(s):
+        # Placeholder functions
+        s.send_response(200)
+        s.send_header("Content-type", "text/html")
+        s.end_headers()
+
+    def do_GET(s):
+        """Respond to a GET request."""
+        # Placeholder functions
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: hear is where you'd create the envelope, send it in to the agent, wait for the response envelope and then serve the response

File: packages/fetchai/connections/http/connection.py [hunk 24]
```diff
@@ -0,0 +1,359 @@
...
+        s.wfile.write("<html><head><title>Title goes here.</title></head>")
+        s.wfile.write("<body><p>This is a test.</p>")
+        # If someone went to "http://something.somewhere.net/foo/bar/",
+        # then s.path equals "/foo/bar/".
+        s.wfile.write("<p>You accessed path: %s</p>" % s.path)
+        s.wfile.write("</body></html>")
+
+    def do_POST(s):
+        """Respond to a GET request."""
+        pass
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: here is where you'd create the envelope, sent it to the agent and serve a response (potentially without waiting for response from agent)

File: packages/fetchai/connections/http/connection.py [hunk 25]
```diff
@@ -0,0 +1,359 @@
...
+        self.thread = Thread(target=self.receiving_loop)
+        self.lock = threading.Lock()
+        self.stopped = True
+        # Initializing the internal HTTP server
+        try:
+            self.server = HTTPServer((self.host, self.port), HTTPHandler)
+        except OSError:
+            logger.error(f"{host}:{port} is already in use, please try another Socket.")
+        # Process API specifications into Envelopes to be sent to InBox
+        self._process_api(api_spec)
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/http/connection.py]: I would use the api_spec to filter for valid paths which the handlers should process

**Comment by DavidMinarsch** [comment]
> **Updates:**
> 
> 1. Fixed the /http/connection.yaml file.
> 2. Updated /http_ex/apis/petstore.yaml file to latest OpenAPI v3.0.2 one found here: http://petstore.swagger.io:8080/.
> 3. Added the /http_ex/apis/petstore_sim.yaml file, which is simplified version of petstore.yaml used for HTTPConntection testing.
> 4. Wrote two functions in /http/connection.py that convert the paths in api_spec into Envelopes each. Although the API responses in JSON format is the Envelope message atm, that may change on how the Envelope receives handles it. The full API path is specified in the uri of the Envelope.
> 5. Added HTTP Server and Handler in connection.py that replaces httpCalls from P2P.
> 6. Other minor changes, including renaming to _host_ and _port_.
> 
> **Next steps:**
> 
> 1. Continue populating the HTTPHandler.
> 2. Begin drafting the _petstoresim_ Skill using the petstore_sim.yaml.
> 3. Write the send/receiving Envelop functions in conntection.py between AEA and skill. This is for when the front end client is making requests.
> 4. Add /agents/http_server in the packages with HTTP Connection, Protocol, and example [petstoresim] Skill bundled together and can be easily fetched.
> 5. Test http_server AEA using CLI tool, such as between two different computers. One for front-end client and other for HTTP server AEA.
> 6. Update the docs.

thoughts:

2. seems a bit excessive, keep it simple stupid :)
3. yes, simple!
4. not quite sure what you're trying to do there and why; see my comments

rennext steps:
1. good idea, see my comments for suggestions
2. I'd put this on hold until the connection is finished.
3. this should happen in the handler mostly imo
4. not a priority imo
5. for now I'd focus on the connection and writing unit tests for it
6. especially for the connection, this would be helpful

**Comment by Kevin-Chen0** [comment]
**Updates:**

1) Completed the do_GET and do_POST functions in HTTPHandler so that they can be pushed to the _process_request() function in HTTP channel.
2) Completed the _process_request() function that packages the request (from HTTPHandler) into Envelope based on DavidM's specs and sent _in_queue_ to the Agent.
3) Successfully tested HTTP Connection (to Envelope part) using a GET and a POST query from pet store, as specified in OpenAPI Specs I/O.

**Next steps:**

1) Complete the _send_response() that sends the response back to the front-end client, given the _out_queue_ Envelope from the Agent. I think that this is the last core part of HTTP Connection.
2) Write a preliminary from of Echo Skill handler that receives the _in_queue_ Envelopes and write the response Envelope _out_queue_, which will finally be handled by _send_response().
3) Updates the docs when the following two steps are complete.

**Comment by DavidMinarsch** [comment]
Cancelling to be able to see the changes on a new PR against develop

**Pull request closed without merging**